### PR TITLE
fixed unexpected behavior where the selected function properties don't get updated

### PR DESCRIPTION
**Pull request by kenfdev**
Signed-off-by: Ken Fukuyama <kenfdev@gmail.com>

## Description

Suppose you had function A and B.
Delete function B and then select function A before B gets deleted from the UI.

When function B disappears from the UI, function A's Status becomes `Not Ready`.
This is because, when the `$scope.functions` updates its property by fetching from `system/functions`, the `$scope.selectedFunction` object changes (even though all the properties are the same, it's a different object.

This means

```
refreshFunction($scope.selectedFunction);
```

isn't updating the newly fetched version of the `selectedFunction`, hence even though it's `Ready`, the UI doesn't get updated.

There may be a better way but for now what I came up of is to update the selected function when the `$scope.functions` length changes.

Plus, I found some anti-patterns and fixed them as well. (if not appropriate, I'll create a different issue and PR)

## How Has This Been Tested?

Chrome

The instructions are explained in the Description.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: gateway/assets/index.html [hunk 1]
```diff
@@ -72,7 +72,7 @@ <h1 style="flex: 1 1 auto;">&nbsp; OpenFaaS Portal</h1>
                 <div flex></div>
             </md-content>
 
-            <md-content flex layout="column" ng-repeat="function in functions" ng-show="function.name == selectedFunction.name">
```
Comment (kenfdev) [inline gateway/assets/index.html]: `ng-show` creates DOM elements and is just hiding it with `css`. If we have tens of functions, the tens of DOM will be created event though only 1 is showing.

File: gateway/assets/script/bootstrap.js [hunk 2]
```diff
@@ -31,19 +31,15 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$timeout', '$md
...
-        $scope.isReady = function(selectedFunction) {
-            return (selectedFunction.ready != undefined && selectedFunction.ready == true);
-        }
-
         $scope.invocation.request = "";
         var fetchFunctionsDelay = 3500;
         var queryFunctionDelay = 2500;
         
-        var fetchFunctionsInterval = setInterval(function() {
+        var fetchFunctionsInterval = $interval(function() {
```
Comment (kenfdev) [inline gateway/assets/script/bootstrap.js]: `setInterval` isn't aware of the AngularJS context so is better to use the `$interval` service prepared by AngularJS.

File: gateway/assets/script/bootstrap.js [hunk 3]
```diff
@@ -169,6 +165,14 @@ app.controller("home", ['$scope', '$log', '$http', '$location', '$timeout', '$md
...
                         if (previousItems.length != response.data.length) {
                             $scope.functions = response.data;
+                            
+                            // update the selected function object because the newly fetched object from the API becomes a different object
+                            var filteredSelectedFunction = $filter('filter')($scope.functions, {name: $scope.selectedFunction.name}, true);
+                            if (filteredSelectedFunction && filteredSelectedFunction.length > 0) {
+                                $scope.selectedFunction = filteredSelectedFunction[0];
+                            } else {
+                                $scope.selectedFunction = undefined;
+                            }
```
Comment (kenfdev) [inline gateway/assets/script/bootstrap.js]: Lines above is the main part of the fix. Where it updates the selectedFunction to the new object.

**Comment by alexellis** [comment]
Do you want to update to use $interval?

File: gateway/assets/index.html (hunk 1 above)
Comment (alexellis) [reply gateway/assets/index.html]: Looks like a good change.

File: gateway/assets/script/bootstrap.js (hunk 3 above)
Comment (alexellis) [reply gateway/assets/script/bootstrap.js]: What does `$filter` do? is it the same effect as a for loop and matching on name?

Comment (kenfdev) [reply gateway/assets/script/bootstrap.js]: Yes. I wanted to use [find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) but it's a relatively new function so didn't want to break older browsers.

I wondered if I should add a simple `find` function and call that, or to use `$filter` but chose to take advantage of AngularJS's built in function.

**Comment by kenfdev** [comment]
> Do you want to update to use $interval?
To avoid future weird bugs, I'd say yes.  Using the native timeout/interval and updating the objects outside of the AngularJS context may lead to properties not getting updated on the UI.
It has no problems right now because, we're using the `$http` service. Which is aware of the AngularJS context.

**Pull request merged**